src/tests/zf_apps/zfsink.c

```diff
@@ -243,15 +244,81 @@ void print_attrs(struct zf_attr* attr)
 }
 
 
-static void monitor()
+static int zf_stats_header_print(struct zf_stack* stack, int layout_sz,
+                                  const zf_stats_layout** zf_stats_layout)
+{
+  int i, nic_cnt;
+  nic_cnt = zf_stats_query_layout(stack, zf_stats_layout, layout_sz);
+
+  pthread_mutex_lock(&printf_mutex);
+  /* only display names for the first NIC and assume others are the same */
+  if( nic_cnt > 1 ) {
+    printf("\n#");
+  }
+  for( i = 0; i < zf_stats_layout[0]->zfsl_fields_num; ++i )
+    printf("  %10s", zf_stats_layout[0]->zfsl_fields[i].zfsfl_name);
+  pthread_mutex_unlock(&printf_mutex);
+
+  return nic_cnt;
+}
+
+
+static void zf_stats_print(struct zf_stack* stack, int reset_stats,
+                           uint8_t* stats_data, int nic_cnt,
+                           const zf_stats_layout** zf_stats_layout)
+{
+  int i, n, n_pad;
+  uint8_t* cur_data = stats_data;
+
+  zf_stats_query(stack, stats_data, zf_stats_layout, nic_cnt, reset_stats);
+
+  pthread_mutex_lock(&printf_mutex);
+  for( n = 0; n < nic_cnt; n++ ) {
+    if( nic_cnt > 1 )
+      printf("\n");
+    for( i = 0; i < zf_stats_layout[n]->zfsl_fields_num; ++i ) {
+      const zf_stats_field_layout* f = &zf_stats_layout[n]->zfsl_fields[i];
+      n_pad = strlen(f->zfsfl_name);
+      if( n_pad < 10 )
+        n_pad = 10;
+      switch( f->zfsfl_size ) {
+        case sizeof(uint32_t):
+          printf("  %*d", n_pad, *(uint32_t*)(cur_data + f->zfsfl_offset));
+          break;
+        default:
+          printf("  %*s", n_pad, ".");
+      };
+    }
+    cur_data += zf_stats_layout[n]->zfsl_data_size;
+  }
+  pthread_mutex_unlock(&printf_mutex);
+}
+
+
+static void monitor(struct zf_stack* stack)
 {
   uint64_t now_bytes, prev_bytes;
   struct timeval start, end;
   uint64_t prev_pkts, now_pkts;
   int ms, pkt_rate, mbps;
 
+  const zf_stats_layout* zf_stats_layout[4];
```

Comment:
Is there any way we can get rid of hard coding 4? is there a way to deduce this length?

Reply:
I've added a function that will report how many interfaces there are for a given TCPDirect stack. This will return 1 unless it's a bond when it will return the number of underlying interfaces.